File: packages/ios/Sources/CrumbCore/CrumbJavaScriptCrash.swift
import CryptoKit
import Foundation

public enum CrumbJavaScriptCrashKind: String, Codable, Equatable, Sendable {
    case fatalException = "fatal_exception"
    case unhandledRejection = "unhandled_rejection"
}

public enum CrumbJavaScriptCrashSource: String, Codable, Equatable, Sendable {
    case javascript
    case nativeTerminationWrapper = "native_termination_wrapper"
}

public struct CrumbJavaScriptBreadcrumb: Codable, Equatable, Sendable {
    public let timestamp: Date
    public let category: String
    public let message: String

    public init(timestamp: Date, category: String, message: String) {
        self.timestamp = timestamp
        self.category = category
        self.message = message
    }
}

public struct CrumbJavaScriptCrash: Codable, Equatable, Sendable {
    public let kind: CrumbJavaScriptCrashKind
    public let source: CrumbJavaScriptCrashSource
    public let errorType: String
    public let message: String
    public let rawStack: String?
    public let fingerprint: String?
    public let occurredAt: Date
    public let nativeTerminationWrapper: Bool
    public let breadcrumbs: [CrumbJavaScriptBreadcrumb]

    public init(
        kind: CrumbJavaScriptCrashKind,
        source: CrumbJavaScriptCrashSource = .javascript,
        errorType: String,
        message: String,
        rawStack: String? = nil,
        fingerprint: String? = nil,
        occurredAt: Date = Date(),
        nativeTerminationWrapper: Bool = false,
        breadcrumbs: [CrumbJavaScriptBreadcrumb] = []
    ) {
        self.kind = kind
        self.source = source
        self.errorType = errorType
        self.message = message
        self.rawStack = rawStack
        self.fingerprint = fingerprint
        self.occurredAt = occurredAt
        self.nativeTerminationWrapper = nativeTerminationWrapper
        self.breadcrumbs = breadcrumbs
    }
}

package struct CrumbJavaScriptCrashRecord: Codable, Equatable, Sendable {
    package let reportID: String
    package let appVersion: String
    package let nativeBuild: String
    package let bundleVersion: String?
    package let environment: String
    package let applicationName: String?
    package let customContext: [String: String]
    package let policyStatus: String
    package let workspacePolicyVersion: Int?
    package let crash: CrumbJavaScriptCrash

    package init(
        reportID: String,
        settings: CrumbReportSettings,
        crash: CrumbJavaScriptCrash
    ) {
        self.reportID = reportID
        appVersion = settings.release.appVersion
        nativeBuild = settings.release.nativeBuild
        bundleVersion = settings.release.bundleVersion
        environment = settings.environment
        applicationName = settings.application.name
        customContext = settings.customContext
        policyStatus = settings.policyStatus.rawValue
        workspacePolicyVersion = settings.workspacePolicyVersion
        self.crash = crash
    }
}

/// Synchronously persists only the small, sanitized crash handoff. It is
/// separate from the report queue because JavaScript fatal handlers must not
/// wait for an actor or a report envelope to be assembled.
package final class CrumbJavaScriptCrashStore: @unchecked Sendable {
    package static let shared = CrumbJavaScriptCrashStore(rootURL: defaultRootURL())

    package static let maximumRecords = 8
    package static let maximumFileBytes = 262_144
    package static let maximumBreadcrumbs = 32
    package static let maximumBreadcrumbBytes = 16_384

    private let lock = NSLock()
    private let rootURL: URL
    private let fileManager: FileManager
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    package init(rootURL: URL, fileManager: FileManager = .default) {
        self.rootURL = rootURL
        self.fileManager = fileManager
        encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .millisecondsSince1970
        encoder.outputFormatting = [.sortedKeys]
        decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .millisecondsSince1970
    }

    @discardableResult
    package func record(
        _ crash: CrumbJavaScriptCrash,
        settings: CrumbReportSettings
    ) -> Bool {
        guard let normalized = normalize(crash) else { return false }
        return withLock {
            var records = readRecords()
            let incoming = CrumbJavaScriptCrashRecord(
                reportID: CrumbReportEnvelopeBuilder.makeReportID(),
                settings: settings,
                crash: normalized
            )
            if let index = records.firstIndex(where: {
                $0.crash.fingerprint == normalized.fingerprint
            }) {
                records[index] = merge(records[index], incoming)
            } else {
                guard records.count < Self.maximumRecords else { return false }
                records.append(incoming)
            }
            return writeRecords(records)
        }
    }

    package func records() -> [CrumbJavaScriptCrashRecord] {
        withLock { readRecords() }
    }

    @discardableResult
    package func remove(reportID: String) -> Bool {
        withLock {
            var records = readRecords()
            let originalCount = records.count
            records.removeAll { $0.reportID == reportID }
            guard records.count != originalCount else { return true }
            return writeRecords(records)
        }
    }

    private func normalize(_ crash: CrumbJavaScriptCrash) -> CrumbJavaScriptCrash? {
        guard crash.occurredAt.timeIntervalSince1970.isFinite else { return nil }
        let errorType = sanitize(crash.errorType, maximumBytes: 128)
        let message = sanitize(crash.message, maximumBytes: 4_096)
        guard !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        let rawStack = crash.rawStack
            .map { sanitize($0, maximumBytes: 16_384, preserveLineBreaks: true) }
            .flatMap { $0.isEmpty ? nil : $0 }

        let normalizedBreadcrumbs = crash.breadcrumbs.compactMap { breadcrumb -> CrumbJavaScriptBreadcrumb? in
            guard breadcrumb.timestamp.timeIntervalSince1970.isFinite else { return nil }
            let category = sanitize(breadcrumb.category, maximumBytes: 64)
            let message = sanitize(breadcrumb.message, maximumBytes: 2_048)
            guard !category.isEmpty, !message.isEmpty else { return nil }
            return CrumbJavaScriptBreadcrumb(
                timestamp: breadcrumb.timestamp,
                category: category,
                message: message
            )
        }
        var breadcrumbBytes = 0
        let boundedBreadcrumbs = normalizedBreadcrumbs.suffix(Self.maximumBreadcrumbs).filter { breadcrumb in
            let bytes = breadcrumb.category.utf8.count + breadcrumb.message.utf8.count + 32
            guard breadcrumbBytes + bytes <= Self.maximumBreadcrumbBytes else { return false }
            breadcrumbBytes += bytes
            return true
        }

        let fingerprint = normalizedFingerprint(
            supplied: crash.fingerprint,
            kind: crash.kind,
            errorType: errorType.isEmpty ? "JavaScriptError" : errorType,
            message: message,
            rawStack: rawStack
        )
        let source = crash.source
        return CrumbJavaScriptCrash(
            kind: crash.kind,
            source: source,
            errorType: errorType.isEmpty ? "JavaScriptError" : errorType,
            message: message,
            rawStack: rawStack,
            fingerprint: fingerprint,
            occurredAt: crash.occurredAt,
            nativeTerminationWrapper: crash.nativeTerminationWrapper || source == .nativeTerminationWrapper,
            breadcrumbs: Array(boundedBreadcrumbs)
        )
    }

    private func merge(
        _ existing: CrumbJavaScriptCrashRecord,
        _ incoming: CrumbJavaScriptCrashRecord
    ) -> CrumbJavaScriptCrashRecord {
        let javascript = if incoming.crash.source == .javascript {
            incoming.crash
        } else if existing.crash.source == .javascript {
            existing.crash
        } else {
            incoming.crash
        }
        let source: CrumbJavaScriptCrashSource = if existing.crash.source == .javascript
            || incoming.crash.source == .javascript {
            .javascript
        } else {
            .nativeTerminationWrapper
        }
        let breadcrumbs = mergeBreadcrumbs(existing.crash.breadcrumbs, incoming.crash.breadcrumbs)
        let crash = CrumbJavaScriptCrash(
            kind: javascript.kind,
            source: source,
            errorType: javascript.errorType,
            message: javascript.message,
            rawStack: javascript.rawStack ?? existing.crash.rawStack ?? incoming.crash.rawStack,
            fingerprint: javascript.fingerprint,
            occurredAt: min(existing.crash.occurredAt, incoming.crash.occurredAt),
            nativeTerminationWrapper: existing.crash.nativeTerminationWrapper
                || incoming.crash.nativeTerminationWrapper
                || source == .nativeTerminationWrapper,
            breadcrumbs: breadcrumbs
        )
        return CrumbJavaScriptCrashRecord(
            reportID: existing.reportID,
            appVersion: existing.appVersion,
            nativeBuild: existing.nativeBuild,
            bundleVersion: existing.bundleVersion,
            environment: existing.environment,
            applicationName: existing.applicationName,
            customContext: existing.customContext,
            policyStatus: existing.policyStatus,
            workspacePolicyVersion: existing.workspacePolicyVersion,
            crash: crash
        )
    }

    private func mergeBreadcrumbs(
        _ first: [CrumbJavaScriptBreadcrumb],
        _ second: [CrumbJavaScriptBreadcrumb]
    ) -> [CrumbJavaScriptBreadcrumb] {
        var result: [CrumbJavaScriptBreadcrumb] = []
        for breadcrumb in (first + second).sorted(by: { $0.timestamp < $1.timestamp }) {
            guard !result.contains(breadcrumb) else { continue }
            let bytes = result.reduce(0) { $0 + $1.category.utf8.count + $1.message.utf8.count + 32 }
                + breadcrumb.category.utf8.count + breadcrumb.message.utf8.count + 32
            guard result.count < Self.maximumBreadcrumbs else { break }
            guard bytes <= Self.maximumBreadcrumbBytes else { continue }
            result.append(breadcrumb)
        }
        return result
    }

    private func normalizedFingerprint(
        supplied: String?,
        kind: CrumbJavaScriptCrashKind,
        errorType: String,
        message: String,
        rawStack: String?
    ) -> String {
        if let supplied,
           supplied.range(of: #"^[A-Za-z0-9._-]{1,128}$"#, options: .regularExpression) != nil {
            return supplied
        }
        let value = "\(kind.rawValue)|\(errorType)|\(message)|\(rawStack ?? "")"
        let digest = SHA256.hash(data: Data(value.utf8))
        return "js_" + digest.map { String(format: "%02x", $0) }.joined()
    }

    private func readRecords() -> [CrumbJavaScriptCrashRecord] {
        guard let data = try? Data(contentsOf: fileURL), data.count <= Self.maximumFileBytes else {
            return []
        }
        return (try? decoder.decode([CrumbJavaScriptCrashRecord].self, from: data)) ?? []
    }

    private func writeRecords(_ records: [CrumbJavaScriptCrashRecord]) -> Bool {
        do {
            try fileManager.createDirectory(at: rootURL, withIntermediateDirectories: true)
            try applyStorageAttributes(to: rootURL)
            let data = try encoder.encode(records)
            guard data.count <= Self.maximumFileBytes else { return false }
            try data.write(to: fileURL, options: [.atomic])
            try applyStorageAttributes(to: fileURL)
            return true
        } catch {
            return false
        }
    }

    private var fileURL: URL {
        rootURL.appendingPathComponent("pending-javascript-crashes.json")
    }

    private func sanitize(
        _ value: String,
        maximumBytes: Int,
        preserveLineBreaks: Bool = false
    ) -> String {
        var sanitized = value
        let replacements = [
            (#"(?i)(https?://)[^/\s:@]+:[^/@\s]+@"#, "$1[REDACTED]@"),
            (#"(?i)\bBearer\s+[A-Za-z0-9._~+/=-]+"#, "Bearer [REDACTED]"),
            (
                #"(?i)\b(authorization|cookie|set-cookie|password|passwd|secret|token|api[_-]?key)\s*[:=]\s*(\"[^\"]*\"|'[^']*'|[^\s,;]+)"#,
                "$1=[REDACTED]"
            ),
            (#"(?i)\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b"#, "[REDACTED_EMAIL]"),
            (#"\b(?:\d[ -]*?){13,19}\b"#, "[REDACTED_NUMBER]"),
            (#"([?&][A-Za-z0-9._~-]+)=([^&#\s]*)"#, "$1=[REDACTED]")
        ]
        for (pattern, replacement) in replacements {
            sanitized = sanitized.replacingOccurrences(
                of: pattern,
                with: replacement,
                options: .regularExpression
            )
        }
        sanitized = sanitized.unicodeScalars.reduce(into: "") { result, scalar in
            if preserveLineBreaks && (scalar == "\n" || scalar == "\r") {
                result.append(Character(scalar))
            } else if CharacterSet.controlCharacters.contains(scalar) || scalar == "\u{7f}" {
                result.append(" ")
            } else {
                result.append(Character(scalar))
            }
        }
        return truncateUTF8(sanitized.trimmingCharacters(in: .whitespacesAndNewlines), maximumBytes: maximumBytes)
    }

    private func truncateUTF8(_ value: String, maximumBytes: Int) -> String {
        guard value.utf8.count > maximumBytes else { return value }
        var result = ""
        for character in value {
            let candidate = result + String(character)
            if candidate.utf8.count + 3 > maximumBytes { break }
            result = candidate
        }
        return result + "…"
    }

    private func withLock<Result>(_ action: () -> Result) -> Result {
        lock.lock()
        defer { lock.unlock() }
        return action()
    }

    private func applyStorageAttributes(to url: URL) throws {
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        var mutableURL = url
        try mutableURL.setResourceValues(values)
        #if os(iOS) || os(tvOS) || os(watchOS) || os(visionOS)
        try fileManager.setAttributes(
            [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication],
            ofItemAtPath: url.path
        )
        #endif
    }

    private static func defaultRootURL() -> URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("dev.crumb", isDirectory: true)
    }
}

private extension CrumbJavaScriptCrashRecord {
    init(
        reportID: String,
        appVersion: String,
        nativeBuild: String,
        bundleVersion: String?,
        environment: String,
        applicationName: String?,
        customContext: [String: String],
        policyStatus: String,
        workspacePolicyVersion: Int?,
        crash: CrumbJavaScriptCrash
    ) {
        self.reportID = reportID
        self.appVersion = appVersion
        self.nativeBuild = nativeBuild
        self.bundleVersion = bundleVersion
        self.environment = environment
        self.applicationName = applicationName
        self.customContext = customContext
        self.policyStatus = policyStatus
        self.workspacePolicyVersion = workspacePolicyVersion
        self.crash = crash
    }
}

File: packages/android/crumb-core/src/main/kotlin/dev/crumb/core/CrumbJavaScriptCrash.kt
package dev.crumb.core

import android.content.Context
import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.Locale
import java.util.UUID
import org.json.JSONArray
import org.json.JSONObject

enum class CrumbJavaScriptCrashKind(val wireValue: String) {
    FATAL_EXCEPTION("fatal_exception"),
    UNHANDLED_REJECTION("unhandled_rejection"),
}

enum class CrumbJavaScriptCrashSource(val wireValue: String) {
    JAVASCRIPT("javascript"),
    NATIVE_TERMINATION_WRAPPER("native_termination_wrapper"),
}

data class CrumbJavaScriptBreadcrumb(
    val timestampMillis: Long,
    val category: String,
    val message: String,
)

data class CrumbJavaScriptCrash(
    val kind: CrumbJavaScriptCrashKind,
    val source: CrumbJavaScriptCrashSource = CrumbJavaScriptCrashSource.JAVASCRIPT,
    val errorType: String,
    val message: String,
    val rawStack: String? = null,
    val fingerprint: String? = null,
    val occurredAtMillis: Long = System.currentTimeMillis(),
    val nativeTerminationWrapper: Boolean = false,
    val breadcrumbs: List<CrumbJavaScriptBreadcrumb> = emptyList(),
)

internal data class CrumbJavaScriptCrashRecord(
    val reportId: String,
    val appVersion: String,
    val nativeBuild: String,
    val bundleVersion: String?,
    val environment: String,
    val applicationName: String?,
    val customContext: Map<String, String>,
    val policyStatus: String,
    val workspacePolicyVersion: Int?,
    val crash: CrumbJavaScriptCrash,
)

internal data class CrumbJavaScriptCrashRuntime(
    val osVersion: String,
    val deviceFamily: String,
    val locale: String,
    val timezone: String,
    val processName: String,
    val processId: Int,
)

internal class CrumbJavaScriptCrashStore internal constructor(
    private val root: File,
) {
    private val lock = Any()
    private val file = File(root, PENDING_FILE)

    fun record(crash: CrumbJavaScriptCrash, settings: CrumbReportSettings): Boolean {
        val normalized = CrumbJavaScriptCrashSanitizer.normalize(crash) ?: return false
        return synchronized(lock) {
            val records = readRecords().toMutableList()
            val incoming = CrumbJavaScriptCrashRecord(
                reportId = CrumbReportEnvelopeBuilder.makeReportId(),
                appVersion = settings.release.appVersion,
                nativeBuild = settings.release.nativeBuild,
                bundleVersion = settings.release.bundleVersion,
                environment = settings.environment,
                applicationName = settings.application.name,
                customContext = settings.customContext,
                policyStatus = settings.policyStatus.wireValue,
                workspacePolicyVersion = settings.workspacePolicyVersion,
                crash = normalized,
            )
            val index = records.indexOfFirst { it.crash.fingerprint == normalized.fingerprint }
            if (index >= 0) {
                records[index] = merge(records[index], incoming)
            } else {
                if (records.size >= MAXIMUM_RECORDS) return@synchronized false
                records += incoming
            }
            writeRecords(records)
        }
    }

    fun records(): List<CrumbJavaScriptCrashRecord> = synchronized(lock) { readRecords() }

    fun remove(reportId: String): Boolean = synchronized(lock) {
        val records = readRecords().toMutableList()
        val changed = records.removeAll { it.reportId == reportId }
        if (!changed) return@synchronized true
        writeRecords(records)
    }

    private fun merge(
        existing: CrumbJavaScriptCrashRecord,
        incoming: CrumbJavaScriptCrashRecord,
    ): CrumbJavaScriptCrashRecord {
        val javascript = when {
            incoming.crash.source == CrumbJavaScriptCrashSource.JAVASCRIPT -> incoming.crash
            existing.crash.source == CrumbJavaScriptCrashSource.JAVASCRIPT -> existing.crash
            else -> incoming.crash
        }
        val source = if (
            existing.crash.source == CrumbJavaScriptCrashSource.JAVASCRIPT ||
            incoming.crash.source == CrumbJavaScriptCrashSource.JAVASCRIPT
        ) {
            CrumbJavaScriptCrashSource.JAVASCRIPT
        } else {
            CrumbJavaScriptCrashSource.NATIVE_TERMINATION_WRAPPER
        }
        return existing.copy(
            crash = CrumbJavaScriptCrash(
                kind = javascript.kind,
                source = source,
                errorType = javascript.errorType,
                message = javascript.message,
                rawStack = javascript.rawStack ?: existing.crash.rawStack ?: incoming.crash.rawStack,
                fingerprint = javascript.fingerprint,
                occurredAtMillis = minOf(existing.crash.occurredAtMillis, incoming.crash.occurredAtMillis),
                nativeTerminationWrapper = existing.crash.nativeTerminationWrapper ||
                    incoming.crash.nativeTerminationWrapper ||
                    source == CrumbJavaScriptCrashSource.NATIVE_TERMINATION_WRAPPER,
                breadcrumbs = mergeBreadcrumbs(existing.crash.breadcrumbs, incoming.crash.breadcrumbs),
            ),
        )
    }

    private fun mergeBreadcrumbs(
        first: List<CrumbJavaScriptBreadcrumb>,
        second: List<CrumbJavaScriptBreadcrumb>,
    ): List<CrumbJavaScriptBreadcrumb> {
        val result = mutableListOf<CrumbJavaScriptBreadcrumb>()
        (first + second).sortedBy(CrumbJavaScriptBreadcrumb::timestampMillis).forEach { breadcrumb ->
            if (breadcrumb in result || result.size >= MAXIMUM_BREADCRUMBS) return@forEach
            val bytes = result.sumOf { crumbUtf8Bytes(it.category) + crumbUtf8Bytes(it.message) + 32 } +
                crumbUtf8Bytes(breadcrumb.category) + crumbUtf8Bytes(breadcrumb.message) + 32
            if (bytes <= MAXIMUM_BREADCRUMB_BYTES) result += breadcrumb
        }
        return result
    }

    private fun readRecords(): List<CrumbJavaScriptCrashRecord> {
        if (!file.isFile || file.length() > MAXIMUM_FILE_BYTES) return emptyList()
        return runCatching {
            val array = JSONArray(file.readText(Charsets.UTF_8))
            (0 until array.length()).mapNotNull { index ->
                array.optJSONObject(index)?.toRecord()
            }
        }.getOrDefault(emptyList())
    }

    private fun writeRecords(records: List<CrumbJavaScriptCrashRecord>): Boolean {
        var temporary: File? = null
        return runCatching {
            if (!root.exists() && !root.mkdirs()) error("could not create crash directory")
            if (!root.isDirectory) error("crash path is not a directory")
            val bytes = records.toJson().toString().toByteArray(Charsets.UTF_8)
            if (bytes.size > MAXIMUM_FILE_BYTES) return@runCatching false
            val target = File(root, "$TEMPORARY_PREFIX${UUID.randomUUID()}")
            temporary = target
            FileOutputStream(target).use { stream ->
                stream.write(bytes)
                stream.fd.sync()
            }
            try {
                Files.move(
                    target.toPath(),
                    file.toPath(),
                    StandardCopyOption.ATOMIC_MOVE,
                    StandardCopyOption.REPLACE_EXISTING,
                )
            } catch (_: java.nio.file.AtomicMoveNotSupportedException) {
                Files.move(
                    target.toPath(),
                    file.toPath(),
                    StandardCopyOption.REPLACE_EXISTING,
                )
            }
            true
        }.getOrElse {
            temporary?.delete()
            false
        }
    }

    private fun CrumbJavaScriptCrashRecord.toJson(): JSONObject = JSONObject().apply {
        put("report_id", reportId)
        put("app_version", appVersion)
        put("native_build", nativeBuild)
        putOptional("bundle_version", bundleVersion)
        put("environment", environment)
        putOptional("application_name", applicationName)
        put("custom_context", JSONObject().apply {
            customContext.forEach { (key, value) -> put(key, value) }
        })
        put("policy_status", policyStatus)
        putOptional("policy_version", workspacePolicyVersion)
        put("crash", crash.toJson())
    }

    private fun CrumbJavaScriptCrash.toJson(): JSONObject = JSONObject().apply {
        put("kind", kind.wireValue)
        put("source", source.wireValue)
        put("error_type", errorType)
        put("message", message)
        putOptional("raw_stack", rawStack)
        putOptional("fingerprint", fingerprint)
        put("occurred_at_millis", occurredAtMillis)
        put("native_termination_wrapper", nativeTerminationWrapper)
        put("breadcrumbs", JSONArray().apply {
            breadcrumbs.forEach { breadcrumb ->
                put(JSONObject().apply {
                    put("timestamp_millis", breadcrumb.timestampMillis)
                    put("category", breadcrumb.category)
                    put("message", breadcrumb.message)
                })
            }
        })
    }

    private fun JSONObject.toRecord(): CrumbJavaScriptCrashRecord? = runCatching {
        val crashObject = getJSONObject("crash")
        val kind = CrumbJavaScriptCrashKind.entries.first { it.wireValue == crashObject.getString("kind") }
        val source = CrumbJavaScriptCrashSource.entries.first { it.wireValue == crashObject.getString("source") }
        val breadcrumbs = crashObject.optJSONArray("breadcrumbs")?.let { array ->
            (0 until array.length()).mapNotNull { index ->
                array.optJSONObject(index)?.let { breadcrumb ->
                    CrumbJavaScriptBreadcrumb(
                        timestampMillis = breadcrumb.getLong("timestamp_millis"),
                        category = breadcrumb.getString("category"),
                        message = breadcrumb.getString("message"),
                    )
                }
            }
        }.orEmpty()
        CrumbJavaScriptCrashRecord(
            reportId = getString("report_id"),
            appVersion = getString("app_version"),
            nativeBuild = getString("native_build"),
            bundleVersion = optionalString("bundle_version"),
            environment = getString("environment"),
            applicationName = optionalString("application_name"),
            customContext = optJSONObject("custom_context")?.let { context ->
                context.keys().asSequence().associateWith(context::getString)
            }.orEmpty(),
            policyStatus = getString("policy_status"),
            workspacePolicyVersion = if (has("policy_version") && !isNull("policy_version")) {
                getInt("policy_version")
            } else {
                null
            },
            crash = CrumbJavaScriptCrash(
                kind = kind,
                source = source,
                errorType = crashObject.getString("error_type"),
                message = crashObject.getString("message"),
                rawStack = crashObject.optionalString("raw_stack"),
                fingerprint = crashObject.optionalString("fingerprint"),
                occurredAtMillis = crashObject.getLong("occurred_at_millis"),
                nativeTerminationWrapper = crashObject.optBoolean("native_termination_wrapper"),
                breadcrumbs = breadcrumbs,
            ),
        )
    }.getOrNull()

    private fun List<CrumbJavaScriptCrashRecord>.toJson(): JSONArray = JSONArray().apply {
        forEach { put(it.toJson()) }
    }

    private fun JSONObject.putOptional(name: String, value: Any?) {
        if (value != null) put(name, value)
    }

    private fun JSONObject.optionalString(name: String): String? =
        if (has(name) && !isNull(name)) getString(name) else null

    companion object {
        private const val PENDING_FILE = "pending-javascript-crashes.json"
        private const val TEMPORARY_PREFIX = ".pending-javascript-crash-"
        private const val MAXIMUM_RECORDS = 8
        private const val MAXIMUM_FILE_BYTES = 262_144L
        private const val MAXIMUM_BREADCRUMBS = 32
        private const val MAXIMUM_BREADCRUMB_BYTES = 16_384

        fun from(context: Context): CrumbJavaScriptCrashStore = CrumbJavaScriptCrashStore(
            File(File(context.applicationContext.noBackupFilesDir, "dev.crumb"), "javascript")
        )
    }
}

internal object CrumbJavaScriptCrashSanitizer {
    private const val MAXIMUM_ERROR_TYPE_BYTES = 128
    private const val MAXIMUM_MESSAGE_BYTES = 4_096
    private const val MAXIMUM_STACK_BYTES = 16_384
    private const val MAXIMUM_BREADCRUMB_CATEGORY_BYTES = 64
    private const val MAXIMUM_BREADCRUMB_MESSAGE_BYTES = 2_048
    private const val MAXIMUM_BREADCRUMBS = 32
    private const val MAXIMUM_BREADCRUMB_BYTES = 16_384
    private val replacements = listOf(
        Regex("(?i)(https?://)[^/\\s:@]+:[^/@\\s]+@") to "$1[REDACTED]@",
        Regex("(?i)\\bBearer\\s+[A-Za-z0-9._~+/=-]+") to "Bearer [REDACTED]",
        Regex(
            "(?i)\\b(authorization|cookie|set-cookie|password|passwd|secret|token|api[_-]?key)" +
                "\\s*[:=]\\s*(\\\"[^\\\"]*\\\"|'[^']*'|[^\\s,;]+)",
        ) to "$1=[REDACTED]",
        Regex("(?i)\\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,}\\b") to "[REDACTED_EMAIL]",
        Regex("\\b(?:\\d[ -]*?){13,19}\\b") to "[REDACTED_NUMBER]",
        Regex("([?&][A-Za-z0-9._~-]+)=([^&#\\s]*)") to "$1=[REDACTED]",
    )

    fun normalize(crash: CrumbJavaScriptCrash): CrumbJavaScriptCrash? {
        if (crash.occurredAtMillis < 0) return null
        val errorType = sanitize(crash.errorType, MAXIMUM_ERROR_TYPE_BYTES)
            .ifBlank { "JavaScriptError" }
        val message = sanitize(crash.message, MAXIMUM_MESSAGE_BYTES)
        if (message.isBlank()) return null
        val rawStack = crash.rawStack?.let { sanitize(it, MAXIMUM_STACK_BYTES, true) }
            ?.takeIf(String::isNotBlank)
        val breadcrumbs = crash.breadcrumbs.mapNotNull { breadcrumb ->
            if (breadcrumb.timestampMillis < 0) return@mapNotNull null
            val category = sanitize(breadcrumb.category, MAXIMUM_BREADCRUMB_CATEGORY_BYTES)
            val message = sanitize(breadcrumb.message, MAXIMUM_BREADCRUMB_MESSAGE_BYTES)
            if (category.isBlank() || message.isBlank()) null else {
                CrumbJavaScriptBreadcrumb(breadcrumb.timestampMillis, category, message)
            }
        }.takeLast(MAXIMUM_BREADCRUMBS).let { values ->
            val result = mutableListOf<CrumbJavaScriptBreadcrumb>()
            var bytes = 0
            values.forEach { breadcrumb ->
                val size = crumbUtf8Bytes(breadcrumb.category) + crumbUtf8Bytes(breadcrumb.message) + 32
                if (bytes + size <= MAXIMUM_BREADCRUMB_BYTES) {
                    result += breadcrumb
                    bytes += size
                }
            }
            result
        }
        val fingerprint = if (crash.fingerprint?.matches(IDENTIFIER_PATTERN) == true) {
            crash.fingerprint
        } else {
            "js_" + CrumbPolicy.sha256(
                listOf(crash.kind.wireValue, errorType, message, rawStack.orEmpty())
                    .joinToString("|")
            )
        }
        return CrumbJavaScriptCrash(
            kind = crash.kind,
            source = crash.source,
            errorType = errorType,
            message = message,
            rawStack = rawStack,
            fingerprint = fingerprint,
            occurredAtMillis = crash.occurredAtMillis,
            nativeTerminationWrapper = crash.nativeTerminationWrapper ||
                crash.source == CrumbJavaScriptCrashSource.NATIVE_TERMINATION_WRAPPER,
            breadcrumbs = breadcrumbs,
        )
    }

    fun truncateUtf8(value: String, maximumBytes: Int): String {
        if (crumbUtf8Bytes(value) <= maximumBytes) return value
        val result = StringBuilder()
        value.forEach { character ->
            val candidate = result.toString() + character
            if (crumbUtf8Bytes(candidate) + 3 > maximumBytes) return@forEach
            result.append(character)
        }
        return "$result…"
    }

    private fun sanitize(value: String, maximumBytes: Int, preserveLineBreaks: Boolean = false): String {
        val replaced = replacements.fold(value) { result, (pattern, replacement) ->
            result.replace(pattern, replacement)
        }
        val printable = replaced.map { character ->
            if (preserveLineBreaks && (character == '\n' || character == '\r')) {
                character
            } else if (character.isISOControl()) {
                ' '
            } else {
                character
            }
        }.joinToString("").trim()
        return truncateUtf8(printable, maximumBytes)
    }

    private val IDENTIFIER_PATTERN = Regex("^[A-Za-z0-9._-]{1,128}$")
}

private fun crumbUtf8Bytes(value: String): Int = value.toByteArray(Charsets.UTF_8).size
